### ISSUE TITLE
vmctl: Allow to set custom retry limit, backoff delay and backoff factor

### ISSUE THREAD

**Issue by nemobis** [VictoriaMetrics/VictoriaMetrics#6622]
### Is your feature request related to a problem? Please describe

I'm trying to migrate 2.5T datapoints (1.65 TiB) from VictoriaMetrics single to cluster with vmctl native. My attempts invariably fail with:
```
migration failed: export/import error: failed to migrate from <redacted>/api/v1/export/native to http://victoriametrics-victoria-metrics-cluster-vminsert.victoriametrics.svc.cluster.local:8480/insert/0/prometheus/api/v1/import/native (retry attempts: 10): execution failed after 10 retry attempts
```
When this happens, usually I/O is saturated on the source side and/or vminsert concurrency is saturated on the receiving end.

### Describe the solution you'd like

I'm fine with the transfer taking longer, as long as it doesn't fail, so I need more retries and a much higher backoff factor. I'd probably want the first retry to wait at least as long as the query timeout (30 seconds or more), and it would be ok to pause for many minutes or even hours if for whatever reason there's slowness.

In https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4447, backoff was increased compared to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4997#issuecomment-1728363055, but not enough. Only the 5th retry goes above 30 seconds, so on the 5th retry there might be up to 5 ongoing requests, multiplied by the concurrency. A single very slow query may slow down all the others, ensuring that they eventually fail. Failure is guaranteed also if for whatever reason resources are exhausted for 20 minutes, as there are no retries beyond such a period.

### Describe alternatives you've considered

* I can reduce concurrency further (I tried 16 and 8, now 4; on 16-core nodes), but this would limit my average throughput to the worst case scenario.
* I can try a lower `--vm-rate-limit` to reduce spikes (I started from 300 MB/s, now trying 150) but that won't help with the export side.
* I could use "minute" for `--vm-native-step-interval` (currently using hour) but then the migration would take months instead of weeks.
* I could disable retries with `--vm-native-disable-retries` (as suggested in https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5366#issuecomment-1827616268), if that meant skipping the high-cardinality metrics which time out on export, but I don't want to skip ingestion of other metrics which just happen to be migrated around the same time as a heavy one. 
* I can't increase I/O speed on the source side because that would require a disk migration.
* I could increase `dedup.minScrapeInterval` to 5 minutes to reduce the amount of data exported ([seeing that](https://docs.victoriametrics.com/single-server-victoriametrics/#how-to-export-data-in-json-line-format) it is «applied to the data exported via /api/v1/export by default»), but I'm not super sure what the result would be for our typical queries.

### Additional information

Source is VictoriaMetrics v1.96, target v1.101.

I'm using `-search.maxUniqueTimeseries=20000000` on the source side per https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6622

**Comment by nemobis** [comment]
This is ingestion rate with concurrency 8, rate of 300 MB/s:

![Kuvakaappaus_20240710_110510](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/901528/6ab0afae-a158-4710-96e9-341e71556548)

This with concurrency 6, rate 200 MB/s:

![Kuvakaappaus_20240710_110442](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/901528/e88b5946-2839-4396-808f-fd0dd88a3d5f)

The latest attempt failed over this metric:
```
Requests to make: 580282 / 19656456 [█████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒] 2.95%
2024/07/10 05:48:14 migration failed: export/import error: failed to migrate from <redacted>/api/v1/export/native to http://victoriametrics-victoria-metrics-cluster-vminsert.victoriametrics.svc.cluster.local:8480/insert/0/prometheus/api/v1/import/native (retry attempts: 10): execution failed after 10 retry attempts
with filter
        filter: match[]={__name__="node_schedstat_timeslices_total"}
        start: 2023-07-09T17:00:00Z
        end: 2023-07-09T18:00:00Z
```

This happened to always time out on export because the query took over 30 seconds. After increasing the timeout on the source side we can see that there are about 50k timeseries for that metric:

![Kuvakaappaus_20240710_104206](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/901528/d698cc26-54d0-4dc7-bc7c-aad448c8bb0c)

On the same minute of the failure, source I/O was saturated:

![Kuvakaappaus_20240710_102034](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/901528/84a86e72-b44a-49aa-af3d-f4e20e0a7dad)

Concurrent inserts on the destination too:

![Kuvakaappaus_20240710_100635](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/901528/2d8e65c6-5f05-403c-a3ab-1b12b4a3d9a5)

But the underlying node for the destination was not actually running out of resources and in fact resource usage was trending down for 15 minutes when the failure happened:

![Kuvakaappaus_20240710_112540](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/901528/63cfb876-c567-4dcd-b5b5-6f6af714d873)

**Comment by nemobis** [comment]
For what it's worth, in my case reducing the concurrency (to 4) and the

**Comment by dmitryk-dk** [comment]
Hi @nemobis ! As far as I can see from your last comment, the problem was with query timeout. Sometimes, when single version of VictoriaMetrics or cluster version has a huge amount of data it is nesessary to change flags like 

`search.maxUniqueTimeseries: "25000000"`

`search.maxQueryDuration: "600s"`
Those values are just an example. But in most cases, it helps to wait to process queries with a huge amount of data.

```-search.maxUniqueTimeseries limits the number of unique time series a single query can find and process. VictoriaMetrics keeps in memory some metainformation about the time series located by each query and spends some CPU time for processing the found time series. This means that the maximum memory usage and CPU usage a single query can use is proportional to -search.maxUniqueTimeseries.```

```-search.maxQueryDuration limits the duration of a single query. If the query takes longer than the given duration, then it is canceled. This allows saving CPU and RAM when executing unexpected heavy queries.```

I am not sure that an increase of the backoff policy is needed

**Comment by nemobis** [comment]
@dmitryk-dk Thanks. As I see it, slow queries which timed out were the root issue, but what caused vmctl to abort was the _buildup_ in queries, which would not happen with a more conservative backoff. So I would still recommend defaults which make it harder to shoot yourself in the foot, but maybe my use case is unusual. :)

**Comment by dmitryk-dk** [comment]
> @dmitryk-dk Thanks. As I see it, slow queries which timed out were the root issue, but what caused vmctl to abort was the _buildup_ in queries, which would not happen with a more conservative backoff. So I would still recommend defaults which make it harder to shoot yourself in the foot, but maybe my use case is unusual. :)

Maybe it is a good idea to use some flags to configure the backoff policy. We had thoughts about it earlier but decided that it was not needed because, in the most cases, the problem was with complicated queries that were not finished and stopped by timeout. 
@hagen1778 wdyt about adding flags to configure backoff policy?

**Comment by nemobis** [comment]
I'm now trying higher hardcoded settings like https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6634 (as I had to restart anyway due to an unrelated issue, i.e. I had to increase `--search.maxSamplesPerQuery` tenfold to export another high cardinality metric). I opened the PR just for illustrative purposes, feel free to close it.

**Comment by hagen1778** [comment]
> @hagen1778 wdyt about adding flags to configure backoff policy?

Yes. Seems like it is better than making users to re-compile the binary. Do you want to take a look @dmitryk-dk ?

@nemobis, thanks for such a detailed investigation. It was expected that default backoff settings would work for most of the cases and extra config-flags will just contribute to configuration complexity.

**Comment by dmitryk-dk** [comment]
> > @hagen1778 wdyt about adding flags to configure backoff policy?
> 
> Yes. Seems like it is better than making users to re-compile the binary. Do you want to take a look @dmitryk-dk ?
> 
> @nemobis, thanks for such a detailed investigation. It was expected that default backoff settings would work for most of the cases and extra config-flags will just contribute to configuration complexity.

Yes , I will make changes soon. Thank you :)

**Comment by hagen1778** [comment]
https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6639 adds `--vm-backoff-retries`, `--vm-backoff-factor`, `--vm-backoff-min-duration` and `--vm-native-backoff-retries`, `--vm-native-backoff-factor`, `--vm-native-backoff-min-duration` command-line flags. This change will be included into the next release.

**Comment by hagen1778** [comment]
This feature was included into [v1.103.0](https://docs.victoriametrics.com/changelog/#v11030).
